File: shared/interfaces/core_types.ts
/**
 * Core TypeScript interfaces for the therapeutic gamification app
 * These interfaces define the core entities and data structures used across all services
 */

// Core Enums
export enum ChapterType {
  SELF_DISCIPLINE = "self_discipline",
  EMPATHY = "empathy",
  RESILIENCE = "resilience",
  CURIOSITY = "curiosity",
  COMMUNICATION = "communication",
  CREATIVITY = "creativity",
  COURAGE = "courage",
  WISDOM = "wisdom"
}

export enum CrystalAttribute {
  SELF_DISCIPLINE = "self_discipline",  // 自律
  EMPATHY = "empathy",                  // 共感
  RESILIENCE = "resilience",            // 回復力
  CURIOSITY = "curiosity",              // 好奇心
  COMMUNICATION = "communication",      // コミュニケーション
  CREATIVITY = "creativity",            // 創造性
  COURAGE = "courage",                  // 勇気
  WISDOM = "wisdom"                     // 知恵
}

export enum CrystalGrowthEvent {
  TASK_COMPLETION = "task_completion",
  STORY_CHOICE = "story_choice",
  MOOD_IMPROVEMENT = "mood_improvement",
  REFLECTION_ENTRY = "reflection_entry",
  SOCIAL_INTERACTION = "social_interaction",
  CREATIVE_ACTIVITY = "creative_activity",
  CHALLENGE_OVERCOME = "challenge_overcome",
  WISDOM_GAINED = "wisdom_gained"
}

export enum TaskType {
  ROUTINE = "routine",
  ONE_SHOT = "one_shot",
  SKILL_UP = "skill_up",
  SOCIAL = "social"
}

export enum TaskStatus {
  PENDING = "pending",
  IN_PROGRESS = "in_progress",
  COMPLETED = "completed",
  OVERDUE = "overdue"
}

export enum CellStatus {
  LOCKED = "locked",
  AVAILABLE = "available",
  IN_PROGRESS = "in_progress",
  COMPLETED = "completed"
}

export enum GuardianPermission {
  VIEW_ONLY = "view_only",
  TASK_EDIT = "task_edit",
  CHAT_SEND = "chat_send"
}

export enum NodeType {
  OPENING = "opening",
  CHALLENGE = "challenge",
  CHOICE = "choice",
  RESOLUTION = "resolution",
  REFLECTION = "reflection",
  COMPANION_INTRO = "companion_intro",
  MONSTER_ENCOUNTER = "monster_encounter",
  ENDING = "ending"
}

export enum UnlockConditionType {
  TASK_COMPLETION = "task_completion",
  MOOD_THRESHOLD = "mood_threshold",
  TIME_BASED = "time_based",
  LEVEL_REQUIREMENT = "level_requirement",
  COMPANION_RELATIONSHIP = "companion_relationship",
  ACHIEVEMENT_UNLOCK = "achievement_unlock",
  DIARY_COMPLETION = "diary_completion"
}

export enum ItemRarity {
  COMMON = "common",
  UNCOMMON = "uncommon",
  RARE = "rare",
  EPIC = "epic",
  LEGENDARY = "legendary"
}

export enum ItemType {
  WEAPON = "weapon",
  ARMOR = "armor",
  ACCESSORY = "accessory",
  CONSUMABLE = "consumable",
  MAGIC = "magic"
}

export enum JobClass {
  WARRIOR = "warrior",
  HERO = "hero",
  MAGE = "mage",
  PRIEST = "priest",
  SAGE = "sage",
  NINJA = "ninja",
  // Advanced classes
  PALADIN = "paladin",
  ARCHMAGE = "archmage",
  SHADOW_MASTER = "shadow_master"
}

export enum DemonType {
  PROCRASTINATION_DRAGON = "procrastination_dragon",
  ANXIETY_SHADOW = "anxiety_shadow",
  DEPRESSION_VOID = "depression_void",
  SOCIAL_FEAR_GOBLIN = "social_fear_goblin"
}

// Core Data Models
export interface UserProfile {
  uid: string;
  email: string;
  display_name: string;
  player_level: number;
  yu_level: number;
  total_xp: number;
  crystal_gauges: Record<string, number>; // 8 attributes: 0-100 percentage
  current_chapter: string;
  daily_task_limit: number; // ADHD consideration
  care_points: number;
  guardian_permissions: string[];
  adhd_profile: Record<string, any>;
  therapeutic_goals: string[];
  created_at: Date;
  last_active: Date;
}

export interface StoryState {
  uid: string;
  current_chapter: string;
  current_node: string;
  available_edges: string[];
  story_history: Record<string, any>[];
  last_generation_time?: Date;
  unlocked_chapters: string[];
  unlocked_nodes: string[];
  completed_nodes: string[];
  choice_history: Record<string, any>[];
  companion_relationships: Record<string, number>;
  ending_scores: Record<string, number>;
  story_flags: Record<string, string | number | boolean>;
  last_updated: Date;
}

export interface TaskRecord {
  task_id: string;
  uid: string;
  task_type: TaskType;
  title: string;
  description: string;
  difficulty: number; // 1-5 scale
  status: TaskStatus;
  due_date?: Date;
  completion_time?: Date;
  xp_earned: number;
  mood_at_completion?: number; // 1-5 scale
  linked_story_edge?: string;
  habit_tag?: string;
  adhd_support: Record<string, any>; // Pomodoro settings, reminders
  created_at: Date;
}

export interface GameState {
  player_level: number;
  yu_level: number;
  current_chapter: ChapterType;
  crystal_gauges: Record<ChapterType, number>; // 0-100 percentage
  total_xp: number;
  last_resonance_event?: Date;
}

export interface XPCalculation {
  base_xp: number;
  difficulty_multiplier: number;
  mood_coefficient: number; // 0.8 - 1.2
  adhd_assist_multiplier: number; // 1.0 - 1.3
  final_xp: number;
}

export interface User {
  uid: string;
  email: string;
  display_name: string;
  player_level: number;
  yu_level: number;
  total_xp: number;
  created_at: Date;
  last_active: Date;
  adhd_profile: Record<string, any>;
  therapeutic_goals: string[];
  game_state: GameState;
}

export interface MandalaCell {
  id: string;
  position: [number, number]; // [row, col]
  status: CellStatus;
  task_id?: string;
  unlock_conditions: string[];
  xp_reward: number;
  chapter_type: ChapterType;
}

export interface MandalaGrid {
  chapter_type: ChapterType;
  cells: MandalaCell[][]; // 9x9 grid
  center_value: string; // Core value for this chapter
  completion_percentage: number;
}

export interface Task {
  task_id: string;
  uid: string;
  task_type: TaskType;
  title: string;
  description: string;
  difficulty: number; // 1-5 scale
  status: TaskStatus;
  due_date?: Date;
  mandala_cell_id?: string;
  adhd_support: Record<string, any>; // Pomodoro settings, reminders, etc.
  created_at: Date;
  completed_at?: Date;
}

export interface MoodLog {
  uid: string;
  log_date: Date;
  mood_score: number; // 1-5 scale
  notes?: string;
  context_tags: string[];
  calculated_coefficient: number; // 0.8-1.2 for XP calculation
}

export interface UnlockCondition {
  condition_type: UnlockConditionType;
  parameters: Record<string, string | number>;
  required: boolean;
}

export interface StoryChapter {
  chapter_id: string;
  chapter_type: ChapterType;
  title: string;
  description: string;
  unlock_conditions: UnlockCondition[];
  estimated_completion_time: number; // minutes
  therapeutic_focus: string[];
  created_at: Date;
}

export interface StoryNode {
  node_id: string;
  chapter_id: string; // Links to parent chapter
  node_type: NodeType;
  title: string;
  content: string;
  estimated_read_time: number; // minutes
  therapeutic_tags: string[];
  unlock_conditions: UnlockCondition[];
  companion_effects: Record<string, number>;
  mood_effects: Record<string, number>;
  ending_flags: Record<string, string | number | boolean>;
  created_at: Date;
}

export interface StoryEdge {
  edge_id: string;
  from_node_id: string;
  to_node_id: string;
  choice_text: string;
  real_task_id?: string; // Links to real-world task
  habit_tag?: string; // Links to habit tracking
  probability: number;
  therapeutic_weight: number;
  companion_requirements: Record<string, number>;
  achievement_rewards: string[];
  ending_influence: Record<string, number>;
}

export interface UserStoryState {
  uid: string;
  current_chapter_id: string;
  current_node_id: string;
  unlocked_chapters: string[];
  unlocked_nodes: string[];
  completed_nodes: string[];
  choice_history: Array<Record<string, string | Date>>;
  companion_relationships: Record<string, number>;
  ending_scores: Record<string, number>;
  story_flags: Record<string, string | number | boolean>;
  last_updated: Date;
}

export interface GuardianLink {
  guardian_id: string;
  user_id: string;
  permission_level: GuardianPermission;
  created_at: Date;
  approved_by_user: boolean;
  care_points: number;
}

// RPG System Types
export interface RPGItem {
  item_id: string;
  name: string;
  description: string;
  item_type: ItemType;
  rarity: ItemRarity;
  stat_bonuses: Record<string, number>; // focus, resilience, motivation, etc.
  therapeutic_effect?: string;
  flavor_text: string;
  icon_url?: string;
}

export interface UserInventory {
  uid: string;
  items: RPGItem[];
  equipped_items: Record<string, string | null>; // slot -> item_id
  total_coins: number;
  last_updated: Date;
}

export interface JobProgression {
  uid: string;
  current_job: JobClass;
  job_levels: Record<JobClass, number>;
  unlocked_jobs: JobClass[];
  job_skills: Record<string, number>; // skill_name -> level
  stat_bonuses: Record<string, number>; // from job progression
}

export interface InnerDemon {
  demon_id: string;
  demon_type: DemonType;
  name: string;
  hp: number;
  max_hp: number;
  weaknesses: string[]; // action types that damage this demon
  rewards: Record<string, number | string[]>; // coins, items
  therapeutic_message: string;
  appearance_conditions: string[];
}

export interface BattleState {
  uid: string;
  demon: InnerDemon;
  user_recent_actions: string[];
  battle_log: string[];
  is_active: boolean;
  started_at: Date;
}

// Crystal System Types
export interface CrystalState {
  attribute: CrystalAttribute;
  current_value: number; // 0-100 percentage
  growth_rate: number; // Multiplier for growth events
  last_growth_event?: Date;
  milestone_rewards: string[]; // Unlocked rewards at milestones
  therapeutic_insights: string[]; // Generated insights for this attribute
}

export interface CrystalGrowthRecord {
  uid: string;
  attribute: CrystalAttribute;
  event_type: CrystalGrowthEvent;
  growth_amount: number; // Points added to crystal
  trigger_context: Record<string, any>; // Context that triggered growth
  therapeutic_message?: string;
  created_at: Date;
}

export interface CrystalMilestone {
  attribute: CrystalAttribute;
  threshold: number; // 0-100 percentage
  title: string;
  description: string;
  rewards: string[]; // Item IDs, abilities, etc.
  therapeutic_benefit: string;
  unlock_content: string[]; // Story nodes, features, etc.
}

export interface UserCrystalSystem {
  uid: string;
  crystals: Record<CrystalAttribute, CrystalState>;
  total_growth_events: number;
  resonance_level: number; // Overall harmony between crystals
  last_resonance_check?: Date;
  active_synergies: string[]; // Active crystal combinations
  growth_history: CrystalGrowthRecord[];
}

export interface CrystalSynergy {
  synergy_id: string;
  name: string;
  required_attributes: CrystalAttribute[];
  min_levels: Record<CrystalAttribute, number>; // Minimum levels required
  effect_description: string;
  stat_bonuses: Record<string, number>;
  therapeutic_benefit: string;
  story_unlock?: string;
}

// Growth Note (グルノート) System Types
export interface GrowthNoteEntry {
  uid: string;
  entry_date: Date;
  current_problems: string; // ①現実世界で困っていること
  ideal_world: string; // ②理想的な世界とは
  ideal_emotions: string; // ③理想的な世界に住むあなたの感情は？
  tomorrow_actions: string; // ④明日から何が出来る？
  xp_earned: number;
  emotional_tone?: string;
  key_themes: string[];
  crystal_growth_triggered: CrystalAttribute[]; // Which crystals grew from this entry
  created_at: Date;
}

export interface ReflectionStreak {
  uid: string;
  current_streak: number;
  longest_streak: number;
  total_reflections: number;
  last_reflection_date?: Date;
  missed_days_this_week: number;
  needs_reminder: boolean;
}

// Utility Types
export type ValidationResult = {
  isValid: boolean;
  errors: string[];
  warnings: string[];
};

export type APIResponse<T> = {
  success: boolean;
  data?: T;
  error?: string;
  timestamp: Date;
};

export type PaginatedResponse<T> = APIResponse<{
  items: T[];
  total: number;
  page: number;
  pageSize: number;
  hasNext: boolean;
}>;

// Constants
export const CRYSTAL_ATTRIBUTES = Object.values(CrystalAttribute);
export const TASK_TYPES = Object.values(TaskType);
export const TASK_STATUSES = Object.values(TaskStatus);
export const CHAPTER_TYPES = Object.values(ChapterType);

export const XP_CALCULATION_CONSTANTS = {
  BASE_XP_PER_DIFFICULTY: 10,
  MOOD_COEFFICIENT_RANGE: { min: 0.8, max: 1.2 },
  ADHD_ASSIST_RANGE: { min: 1.0, max: 1.3 },
  RESONANCE_BONUS_MULTIPLIER: 1.5
} as const;

export const MANDALA_CONSTANTS = {
  GRID_SIZE: 9,
  CENTER_POSITION: [4, 4] as [number, number],
  MAX_DAILY_TASKS: 16
} as const;

export const THERAPEUTIC_CONSTANTS = {
  MOOD_SCALE_MIN: 1,
  MOOD_SCALE_MAX: 5,
  CRYSTAL_MAX_VALUE: 100,
  RESONANCE_LEVEL_THRESHOLD: 5,
  GROWTH_NOTE_XP_REWARD: 25
} as const;

File: shared/interfaces/core_types.py
"""
Core TypeScript-style interfaces for the therapeutic gamification app
These interfaces define the core entities and data structures used across all services
"""

from typing import Dict, List, Optional, Union, Literal, Any
from datetime import datetime
from pydantic import BaseModel
from enum import Enum

# Core Enums
class ChapterType(str, Enum):
    SELF_DISCIPLINE = "self_discipline"
    EMPATHY = "empathy"
    RESILIENCE = "resilience"
    CURIOSITY = "curiosity"
    COMMUNICATION = "communication"
    CREATIVITY = "creativity"
    COURAGE = "courage"
    WISDOM = "wisdom"

class CrystalAttribute(str, Enum):
    """8つ"""
    SELF_DISCIPLINE = "self_discipline"  # 自動
    EMPATHY = "empathy"                  # 共有
    RESILIENCE = "resilience"            # ?
    CURIOSITY = "curiosity"              # ?
    COMMUNICATION = "communication"      # コア
    CREATIVITY = "creativity"            # 創
    COURAGE = "courage"                  # 勇
    WISDOM = "wisdom"                    # ?

class CrystalGrowthEvent(str, Enum):
    """?"""
    TASK_COMPLETION = "task_completion"
    STORY_CHOICE = "story_choice"
    MOOD_IMPROVEMENT = "mood_improvement"
    REFLECTION_ENTRY = "reflection_entry"
    SOCIAL_INTERACTION = "social_interaction"
    CREATIVE_ACTIVITY = "creative_activity"
    CHALLENGE_OVERCOME = "challenge_overcome"
    WISDOM_GAINED = "wisdom_gained"

class TaskType(str, Enum):
    ROUTINE = "routine"
    ONE_SHOT = "one_shot"
    SKILL_UP = "skill_up"
    SOCIAL = "social"

class TaskStatus(str, Enum):
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    OVERDUE = "overdue"

class CellStatus(str, Enum):
    LOCKED = "locked"
    AVAILABLE = "available"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"

class GuardianPermission(str, Enum):
    VIEW_ONLY = "view_only"
    TASK_EDIT = "task_edit"
    CHAT_SEND = "chat_send"

# Core Data Models
class UserProfile(BaseModel):
    """Core user profile with comprehensive game state and therapeutic data"""
    uid: str
    email: str
    display_name: str
    player_level: int = 1
    yu_level: int = 1
    total_xp: int = 0
    crystal_gauges: Dict[str, int] = {}  # 8 attributes: 0-100 percentage
    current_chapter: str = "self_discipline"
    daily_task_limit: int = 16  # ADHD consideration
    care_points: int = 0
    guardian_permissions: List[str] = []
    adhd_profile: Dict = {}
    therapeutic_goals: List[str] = []
    created_at: datetime
    last_active: datetime
    
    def __init__(self, **data):
        super().__init__(**data)
        # Initialize crystal gauges if empty
        if not self.crystal_gauges:
            self.crystal_gauges = {
                "Self-Discipline": 0,
                "Empathy": 0,
                "Resilience": 0,
                "Curiosity": 0,
                "Communication": 0,
                "Creativity": 0,
                "Courage": 0,
                "Wisdom": 0
            }

class StoryState(BaseModel):
    """User's story progression state with DAG navigation"""
    uid: str
    current_chapter: str
    current_node: str
    available_edges: List[str] = []
    story_history: List[Dict] = []
    last_generation_time: Optional[datetime] = None
    unlocked_chapters: List[str] = []
    unlocked_nodes: List[str] = []
    completed_nodes: List[str] = []
    choice_history: List[Dict] = []
    companion_relationships: Dict[str, int] = {}
    ending_scores: Dict[str, float] = {}
    story_flags: Dict[str, Union[str, int, float, bool]] = {}
    last_updated: datetime

class TaskRecord(BaseModel):
    """Task record with XP calculation and story integration"""
    task_id: str
    uid: str
    task_type: TaskType
    title: str
    description: str
    difficulty: int  # 1-5 scale
    status: TaskStatus = TaskStatus.PENDING
    due_date: Optional[datetime] = None
    completion_time: Optional[datetime] = None
    xp_earned: int = 0
    mood_at_completion: Optional[int] = None  # 1-5 scale
    linked_story_edge: Optional[str] = None
    habit_tag: Optional[str] = None
    adhd_support: Dict = {}  # Pomodoro settings, reminders
    created_at: datetime
    
    def calculate_xp(self, mood_coefficient: float, adhd_assist: float) -> int:
        """Calculate XP using the formula: difficulty ? mood_coefficient ? adhd_assist"""
        base_xp = self.difficulty * 10
        calculated_xp = int(base_xp * mood_coefficient * adhd_assist)
        return calculated_xp

class GameState(BaseModel):
    """Core game state tracking player and Yu progression"""
    player_level: int
    yu_level: int
    current_chapter: ChapterType
    crystal_gauges: Dict[ChapterType, int]  # 0-100 percentage
    total_xp: int
    last_resonance_event: Optional[datetime]
    
class XPCalculation(BaseModel):
    """XP calculation components"""
    base_xp: int
    difficulty_multiplier: float
    mood_coefficient: float  # 0.8 - 1.2
    adhd_assist_multiplier: float  # 1.0 - 1.3
    final_xp: int

class User(BaseModel):
    """Core user entity"""
    uid: str
    email: str
    display_name: str
    player_level: int = 1
    yu_level: int = 1
    total_xp: int = 0
    created_at: datetime
    last_active: datetime
    adhd_profile: Dict = {}
    therapeutic_goals: List[str] = []
    game_state: GameState

class MandalaCell(BaseModel):
    """Individual cell in the 9x9 Mandala grid"""
    id: str
    position: tuple[int, int]  # (row, col)
    status: CellStatus
    task_id: Optional[str] = None
    unlock_conditions: List[str] = []
    xp_reward: int = 0
    chapter_type: ChapterType

class MandalaGrid(BaseModel):
    """Complete 9x9 Mandala grid for a chapter"""
    chapter_type: ChapterType
    cells: List[List[MandalaCell]]  # 9x9 grid
    center_value: str  # Core value for this chapter
    completion_percentage: float

class Task(BaseModel):
    """Task entity with ADHD support features"""
    task_id: str
    uid: str
    task_type: TaskType
    title: str
    description: str
    difficulty: int  # 1-5 scale
    status: TaskStatus
    due_date: Optional[datetime]
    mandala_cell_id: Optional[str]
    adhd_support: Dict = {}  # Pomodoro settings, reminders, etc.
    created_at: datetime
    completed_at: Optional[datetime]

class MoodLog(BaseModel):
    """Daily mood tracking"""
    uid: str
    log_date: datetime
    mood_score: int  # 1-5 scale
    notes: Optional[str] = ""
    context_tags: List[str] = []
    calculated_coefficient: float  # 0.8-1.2 for XP calculation

class NodeType(str, Enum):
    OPENING = "opening"
    CHALLENGE = "challenge"
    CHOICE = "choice"
    RESOLUTION = "resolution"
    REFLECTION = "reflection"
    COMPANION_INTRO = "companion_intro"
    MONSTER_ENCOUNTER = "monster_encounter"
    ENDING = "ending"

class UnlockConditionType(str, Enum):
    TASK_COMPLETION = "task_completion"
    MOOD_THRESHOLD = "mood_threshold"
    TIME_BASED = "time_based"
    LEVEL_REQUIREMENT = "level_requirement"
    COMPANION_RELATIONSHIP = "companion_relationship"
    ACHIEVEMENT_UNLOCK = "achievement_unlock"
    DIARY_COMPLETION = "diary_completion"

class UnlockCondition(BaseModel):
    condition_type: UnlockConditionType
    parameters: Dict[str, Union[str, int, float]]
    required: bool = True

class StoryChapter(BaseModel):
    """Story chapter containing multiple nodes - CHAPTER level"""
    chapter_id: str
    chapter_type: ChapterType
    title: str
    description: str
    unlock_conditions: List[UnlockCondition] = []
    estimated_completion_time: int  # minutes
    therapeutic_focus: List[str] = []
    created_at: datetime

class StoryNode(BaseModel):
    """Story DAG node - NODE level"""
    node_id: str
    chapter_id: str  # Links to parent chapter
    node_type: NodeType
    title: str
    content: str
    estimated_read_time: int  # minutes
    therapeutic_tags: List[str] = []
    unlock_conditions: List[UnlockCondition] = []
    companion_effects: Dict[str, int] = {}
    mood_effects: Dict[str, float] = {}
    ending_flags: Dict[str, Union[str, int, float, bool]] = {}
    created_at: datetime

class StoryEdge(BaseModel):
    """Story DAG edge connecting nodes - EDGE level"""
    edge_id: str
    from_node_id: str
    to_node_id: str
    choice_text: str
    real_task_id: Optional[str] = None  # Links to real-world task
    habit_tag: Optional[str] = None     # Links to habit tracking
    probability: float = 1.0
    therapeutic_weight: float = 1.0
    companion_requirements: Dict[str, int] = {}
    achievement_rewards: List[str] = []
    ending_influence: Dict[str, float] = {}

class UserStoryState(BaseModel):
    """User's story progression state with comprehensive tracking"""
    uid: str
    current_chapter_id: str
    current_node_id: str
    unlocked_chapters: List[str] = []
    unlocked_nodes: List[str] = []
    completed_nodes: List[str] = []
    choice_history: List[Dict[str, Union[str, datetime]]] = []
    companion_relationships: Dict[str, int] = {}
    ending_scores: Dict[str, float] = {}
    story_flags: Dict[str, Union[str, int, float, bool]] = {}
    last_updated: datetime

class GuardianLink(BaseModel):
    """Guardian-user relationship"""
    guardian_id: str
    user_id: str
    permission_level: GuardianPermission
    created_at: datetime
    approved_by_user: bool = False
    care_points: int = 0

# RPG System Types
class ItemRarity(str, Enum):
    COMMON = "common"
    UNCOMMON = "uncommon"
    RARE = "rare"
    EPIC = "epic"
    LEGENDARY = "legendary"

class ItemType(str, Enum):
    WEAPON = "weapon"
    ARMOR = "armor"
    ACCESSORY = "accessory"
    CONSUMABLE = "consumable"
    MAGIC = "magic"

class JobClass(str, Enum):
    WARRIOR = "warrior"
    HERO = "hero"
    MAGE = "mage"
    PRIEST = "priest"
    SAGE = "sage"
    NINJA = "ninja"
    # Advanced classes
    PALADIN = "paladin"
    ARCHMAGE = "archmage"
    SHADOW_MASTER = "shadow_master"

class DemonType(str, Enum):
    PROCRASTINATION_DRAGON = "procrastination_dragon"
    ANXIETY_SHADOW = "anxiety_shadow"
    DEPRESSION_VOID = "depression_void"
    SOCIAL_FEAR_GOBLIN = "social_fear_goblin"

class RPGItem(BaseModel):
    """RPG item with therapeutic themes"""
    item_id: str
    name: str
    description: str
    item_type: ItemType
    rarity: ItemRarity
    stat_bonuses: Dict[str, int] = {}  # focus, resilience, motivation, etc.
    therapeutic_effect: Optional[str] = None
    flavor_text: str = ""
    icon_url: Optional[str] = None

class UserInventory(BaseModel):
    """User's item inventory"""
    uid: str
    items: List[RPGItem] = []
    equipped_items: Dict[str, Optional[str]] = {}  # slot -> item_id
    total_coins: int = 0
    last_updated: datetime

class JobProgression(BaseModel):
    """User's job class progression"""
    uid: str
    current_job: JobClass
    job_levels: Dict[JobClass, int] = {}
    unlocked_jobs: List[JobClass] = []
    job_skills: Dict[str, int] = {}  # skill_name -> level
    stat_bonuses: Dict[str, int] = {}  # from job progression

class InnerDemon(BaseModel):
    """Inner demon battle entity"""
    demon_id: str
    demon_type: DemonType
    name: str
    hp: int
    max_hp: int
    weaknesses: List[str] = []  # action types that damage this demon
    rewards: Dict[str, Union[int, List[str]]] = {}  # coins, items
    therapeutic_message: str
    appearance_conditions: List[str] = []

class BattleState(BaseModel):
    """Current battle state"""
    uid: str
    demon: InnerDemon
    user_recent_actions: List[str] = []
    battle_log: List[str] = []
    is_active: bool = True
    started_at: datetime

# Crystal System Types
class CrystalState(BaseModel):
    """Individual crystal attribute state"""
    attribute: CrystalAttribute
    current_value: int = 0  # 0-100 percentage
    growth_rate: float = 1.0  # Multiplier for growth events
    last_growth_event: Optional[datetime] = None
    milestone_rewards: List[str] = []  # Unlocked rewards at milestones
    therapeutic_insights: List[str] = []  # Generated insights for this attribute

class CrystalGrowthRecord(BaseModel):
    """Record of crystal growth event"""
    uid: str
    attribute: CrystalAttribute
    event_type: CrystalGrowthEvent
    growth_amount: int  # Points added to crystal
    trigger_context: Dict[str, Any] = {}  # Context that triggered growth
    therapeutic_message: Optional[str] = None
    created_at: datetime

class CrystalMilestone(BaseModel):
    """Crystal milestone definition"""
    attribute: CrystalAttribute
    threshold: int  # 0-100 percentage
    title: str
    description: str
    rewards: List[str] = []  # Item IDs, abilities, etc.
    therapeutic_benefit: str
    unlock_content: List[str] = []  # Story nodes, features, etc.

class UserCrystalSystem(BaseModel):
    """User's complete crystal system state"""
    uid: str
    crystals: Dict[CrystalAttribute, CrystalState] = {}
    total_growth_events: int = 0
    resonance_level: int = 0  # Overall harmony between crystals
    last_resonance_check: Optional[datetime] = None
    active_synergies: List[str] = []  # Active crystal combinations
    growth_history: List[CrystalGrowthRecord] = []
    
    def __init__(self, **data):
        super().__init__(**data)
        # Initialize all crystal states if empty
        if not self.crystals:
            for attr in CrystalAttribute:
                self.crystals[attr] = CrystalState(attribute=attr)

class CrystalSynergy(BaseModel):
    """Crystal synergy combination effects"""
    synergy_id: str
    name: str
    required_attributes: List[CrystalAttribute]
    min_levels: Dict[CrystalAttribute, int]  # Minimum levels required
    effect_description: str
    stat_bonuses: Dict[str, int] = {}
    therapeutic_benefit: str
    story_unlock: Optional[str] = None

# Growth Note (?) System Types
class GrowthNoteEntry(BaseModel):
    """Daily growth note entry"""
    uid: str
    entry_date: datetime
    current_problems: str = ""  # ?
    ideal_world: str = ""       # ?
    ideal_emotions: str = ""    # ?
    tomorrow_actions: str = ""  # ?
    xp_earned: int = 25
    emotional_tone: Optional[str] = None
    key_themes: List[str] = []
    crystal_growth_triggered: List[CrystalAttribute] = []  # Which crystals grew from this entry
    created_at: datetime

class ReflectionStreak(BaseModel):
    """User's reflection streak tracking"""
    uid: str
    current_streak: int = 0
    longest_streak: int = 0
    total_reflections: int = 0
    last_reflection_date: Optional[datetime] = None
    missed_days_this_week: int = 0
    needs_reminder: bool = False